serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/MatrixClient.kt

```diff
@@ -2,82 +2,41 @@ package xyz.room409.serif.serif_shared
 import kotlinx.coroutines.*
 import io.ktor.client.*
 import io.ktor.client.request.*
+import io.ktor.client.statement.*
 import io.ktor.client.features.json.*
 import io.ktor.client.features.json.serializer.KotlinxSerializer
 import io.ktor.http.*
-import kotlinx.serialization.*
-import kotlinx.serialization.json.*
 
 sealed class Outcome<out T : Any> { }
 data class Success<out T : Any>(val value: T) : Outcome<T>()
 data class Error(val message: String, val cause: Exception? = null) : Outcome<Nothing>()
 
-@Serializable
-data class LoginRequest(val type: String, val identifier: LoginIdentifier, val password: String, val initial_device_display_name: String) {
-    constructor(username: String, password: String): this(type="m.login.password",
-                                                          identifier=LoginIdentifier(type="m.id.user",user=username),
-                                                          password=password,
-                                                          initial_device_display_name="Serif")
-}
-@Serializable
-data class LoginIdentifier(val type: String, val user: String)
-
-@Serializable
-data class LoginResponse(val access_token: String)
+// Returns the first Event of class T, or null
+inline fun <reified T> List<Event>.firstOfType(): T? = this.map { (it as? T?) }.firstOrNull { it != null }
 
-@Serializable
-data class RoomMessage(val msgtype: String, val body: String) {
-    constructor(body: String): this(msgtype="m.text",body=body)
-}
-@Serializable
-data class EventIdResponse(val event_id: String)
+class MatrixSession(val client: HttpClient, val access_token: String) {
+    var sync_response: SyncResponse? = null
```

Comment:
Does this have to be null, do we not have some initial/default Sync Response state we can use instead to clean up the !!s down below?

Reply:
Hmm, that's a good point, I should look at that.